flink-runtime/src/main/java/org/apache/flink/runtime/asyncprocessing/ParallelEpochManager.java

```diff
@@ -0,0 +1,41 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.runtime.asyncprocessing;
+
+/** An implementation of {@link AbstractEpochManager} for parallel epoch execution. */
+public class ParallelEpochManager extends AbstractEpochManager {
+
+    public ParallelEpochManager() {
+        super();
+    }
+
+    @Override
+    public void onNonRecord(Runnable action) {
+        Epoch lastEpoch = outputQueue.get(outputQueue.size() - 1);
+        lastEpoch.action = action;
+        lastEpoch.close();
```

Comment:
The action should be invoked when ALL previous epochs have finished, instead of only the last one epoch.

Reply:
Yes, it is, I checked `outputQueue.size() == 1` here.

Reply:
For `Parallel mode`, we don't guarantee that watermark 0 must have been finished before watermark 1.

For `Serial mode`, there is only one epoch in the queue at the same time, which provides the above guarantee as you mentioned.